feat: Placement support

ml019 opened the pull request:
## Intent of Change
- Documentation (new or update to existing)

## Description
Document the decision in regards to completing support for placement of different resource groups in different accounts, as well as handling the question of account level infrastructure.

## Motivation and Context
- better handling of account infrastructure
- better management of multi-region deployments
- less configuration complexity by holding everything in the cmdb

## How Has This Been Tested?
N/A

## Related Changes
<!--- If anything not covered by the headings below, add here  -->

### Prerequisite PRs:
<!--- Add a checklist of items or leave the default of "None" -->
- None

### Dependent PRs:
<!--- Add a checklist of items or leave the default of "None" -->
- None

### Consumer Actions:
<!--- Add a checklist of items or leave the default of "None"
What changes must a consumer of this repository make in order to utilise it?
-->
- None

ml019 committed:
feat: Placement support

Document the decision in regards to completing support for placement of
different resource groups in different accounts, as well as handling
the question of account level infrastructure.

ml019 commented:
@roleyfoley would appreciate some content on `Foundation Components` - I think they represent the only other option I could think of and as your suggestion, you are in the best position to describe them.

The alternative is we just document one option.

roleyfoley commented:
> @roleyfoley would appreciate some content on `Foundation Components` - I think they represent the only other option I could think of and as your suggestion, you are in the best position to describe them.
> 
> The alternative is we just document one option.

Probably better to say that this is an extension/replacement of https://github.com/hamlet-io/architectural-decision-log/blob/main/adr/0002-layer-deployments.md  and that the options considered in that one apply here

ml019 committed:
refactor: build on ADR0002

The ADR is actually just capturing additional design detail for a
decision that was made in ADR0002.

Reflect this in the structure of the ADR.

RossMurr4y commented on adr/0007-placement-support.md:
```suggestion
1. A link must have a `Scope` attribute representing the target-solution's scope. The default, representing existing usage, is the `segment` scope.
```

RossMurr4y commented on adr/0007-placement-support.md:
If a `Subscription` is tied to a `HostingPlatform`, would `HostingSubscription` or `PlatformSubscription` provide an easily identified connection between the two?

ml019 replied:
A subscription is tied to its provider. It is more generic than hosting. It could, for instance, represent an org in github.

This I think `Subscription` is correct. IT might end up with an "Engine" parameter if a provider offers more than one type of subscription but in general I think its likely to be 1:1 with the provider.

RossMurr4y commented on adr/0007-placement-support.md:
I think this could use some clarification here. I understand the concept of `HostingPlatform` broadly, but how do `HostingPlatform`-components differentiate from `Region`?

Does a `HostingPlatform` set a limitation - no component can deploy outside of these locations? 

Or does it represent the capability to deploy to that location? 

Or that realised component instances already exist there?

ml019 replied:
The capability to deploy to that location. I'll add some extra text. I think the idea may evolve as we implement it.

ml019 committed:
refactor: refine wording

Co-authored-by: Ross Murray <ross.murray@gosource.com.au>

ml019 committed:
refactor: add explanation for HostingPlatform

Add some additional explanatory material in regards to the proposed
HostingPlatform component.

roleyfoley commented on adr/0007-placement-support.md:
```suggestion
* links have proven very flexible so expect the proposed changes to be re-purposed for other things in the future
```

roleyfoley commented on adr/0007-placement-support.md:
```suggestion
Currently links are used extensively in product solutions. A link goes from an occurrence to another within a solution. This option builds on existing link processing as follows;
```

I think that sentence is about the current state?

roleyfoley commented on adr/0007-placement-support.md:
The default would actually be self rather than segment right? 
A link configured to be scope segment could be another segment 

So I think here the default would a link within my solution

ml019 replied:
I've updated the text to reflect latest thinking on `districts`. Suggest a reread/review.

roleyfoley commented on adr/0007-placement-support.md:
attributes of the link or is the scope to do with an ordered set of layers?

ml019 replied:
I've updated the text to reflect latest thinking on `districts`. Suggest a reread/review.

roleyfoley commented on adr/0007-placement-support.md:
If we include Tenant here, we might have to consider how names are formed? at the moment tenant isn't included in naming. 

I can see how Tenant fits though. I.e. an organisation that owns the product

ml019 replied:
I've updated the text to reflect latest thinking on `districts`. Suggest a reread/review.

ml019 committed:
fix: typos

Co-authored-by: Michael Foley <roleyfoley@gmail.com>

ml019 committed:
refactor: rephrase explanation of what a link currently is

Co-authored-by: Michael Foley <roleyfoley@gmail.com>

ml019 committed:
refactor: districts

Rename scopes to `districts` and define districts to be an ordered set
of layers.

Refine the sematics of how link processing will work on the basis of a
link having two parts - the solution identifier and the occurrence
identifier. Each of these will be identified by attributes in the link.

ml019 committed:
fix:formatting
